[asan][Darwin] Use Apples blocks extension only when supported (#72639)

tob2 opened the pull request:
Issue #72639

The commit at 020cdaf broke build of asan on macOS with GCC. GCC does not support the Apple blocks extension (yet). Uses of blocks in other parts of the sanitisers are protected by MISSING_BLOCKS_SUPPORT. But the type definition is not.

_This applies FX's patch from the issue._

thesamesam reviewed (approved):
LGTM but I can't formally approve.

fxcoudert committed:
[asan][Darwin] Use Apples blocks extension only when supported (#72639)

The commit at 020cdaf broke build of asan on macOS with GCC. GCC does not
support the Apple blocks extension (yet). Uses of blocks in other parts
of the sanitisers are protected by MISSING_BLOCKS_SUPPORT. But the type
definition is not.

fxcoudert commented:
I can confirm that this patches fixes the build issue, and induces no regression in GCC. Could it be reviewed and merged? 🙏

fxcoudert commented:
Thanks!
